File: Java/QuickSort.java
package Algortithms.Sorting;

import java.util.Arrays;
import java.util.List;

public class QuickSort {
    public Integer [] quickSort(Integer [] array,int left,int right) {
        int partitionIndex;
        int pivotIndex;
        if(left < right){
            System.out.println("partitioning from: "+left+"to: "+right);
           partitionIndex=partition(array, right,left,right);
            System.out.println("quick sort from :"+left+"to: "+(partitionIndex-1));
           quickSort(array,left,partitionIndex-1);
            System.out.println("quick sort from :"+(partitionIndex-1)+"to: "+right);
           quickSort(array,partitionIndex+1,right);
        }

        return array;
    }

//    private int partition(Integer [] array,int pivotIndex,int leftPointer,int rightPointer) { //used partition process
//        while (rightPointer>leftPointer) {
//            while (array[leftPointer] <= array[pivotIndex]) {
//                leftPointer++;
//            }
//
//            while (array[rightPointer] > array[pivotIndex]) {
//                rightPointer--;
//            }
//        if(leftPointer<rightPointer) {
//            swap(array, rightPointer, leftPointer);
//        }
//    }
//        swap(array,pivotIndex,rightPointer);
//        return rightPointer;
//    }

    private int partition(Integer [] array,int pivotIndex,int leftPointer,int rightPointer){
        int pivotValue =array[pivotIndex];
        int pointerSmallElement=leftPointer-1; //this variable holds the latest index value of group of elements which are less than pivot
        for (int i=leftPointer;i<rightPointer;i++){
            if(array[i]<pivotValue){ //if we found a value less than pivot lets swap it (to make it into small element group
                pointerSmallElement++;
                swap(array,i,pointerSmallElement);
            }

        }
        swap(array,rightPointer,pointerSmallElement+1); /*at this point we got our small group and large group..and pivot standing on right..lets swap pivot and
         pointer small element+1 ..so pivot will sit on its appropriate place*/
        return pointerSmallElement+1;
    }
    public void swap(Integer [] array,int index1,int index2){
        System.out.println("swaping "+array[index1]+"and"+array[index2]);
        int temp=array[index1];
        array[index1]=array[index2];
        array[index2]=temp;
    }


    public static void main(String[] args) {
    QuickSort sort = new QuickSort();
    Integer [] array= new Integer[]{10,16,8,2,9,12};
        System.out.println(Arrays.toString(sort.quickSort(array, 0, 5)));
    }
}

File: python/QuickSort.py
#function that takes last element as pivot and places all elements smaller than pivot element to left of pivot and elements
#larger than pvot element to right of pivot
def partition(arr, low, high):
    i = (low-1)         # index of smaller element
    pivot = arr[high]     # pivot
 
    for j in range(low, high):
 
        if arr[j] <= pivot:
 
            i = i+1
            arr[i], arr[j] = arr[j], arr[i]
 
    arr[i+1], arr[high] = arr[high], arr[i+1]
    return (i+1)
 
#Quick Sort function
def quicksort(arr, low, high):
    if len(arr) == 1:
        return arr
    if low < high:
 
        pi = partition(arr, low, high)
 
        quicksort(arr, low, pi-1)
        quicksort(arr, pi+1, high)
 
 
#main program to test code
arr = [5,8,3,4,1,2]
n = len(arr)
quicksort(arr, 0, n-1)
print("Sorted array:")
for i in range(n):
    print(arr[i]),
